dialer.go

```diff
@@ -0,0 +1,247 @@
+
+	// Resolver optionally specifies an alternate resolver to use.
+	Resolver Resolver
+}
+
+// Dial connects to the address on the named network.
+func (d *Dialer) Dial(network string, address string) (*Conn, error) {
+	return d.DialContext(context.Background(), network, address)
+}
+
+// DialContext connects to the address on the named network using the provided
+// context.
+//
+// The provided Context must be non-nil. If the context expires before the
+// connection is complete, an error is returned. Once successfully connected,
+// any expiration of the context will not affect the connection.
+//
+// When using TCP, and the host in the address parameter resolves to multiple
+// network addresses, any dial timeout (from d.Timeout or ctx) is spread over
+// each consecutive dial, such that each is given an appropriate fraction of the
+// time to connect. For example, if a host has 4 IP addresses and the timeout is
+// 1 minute, the connect to each single address will be given 15 seconds to
+// complete before trying the next one.
+func (d *Dialer) DialContext(ctx context.Context, network string, address string) (*Conn, error) {
+	if d.Timeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, d.Timeout)
+		defer cancel()
+	}
+
+	if !d.Deadline.IsZero() {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithDeadline(ctx, d.Deadline)
+		defer cancel()
+	}
+
+	c, err := d.dialContext(ctx, network, address)
+	if err != nil {
+		return nil, err
+	}
+	return NewConnWith(c, ConnConfig{ClientID: d.ClientID}), nil
+}
+
+// DialLeader opens a connection to the leader of the partition for a given
+// topic.
+//
+// The address given to the DialContext method may not be the one that the
+// connection will end up being established to, because the dialer will lookup
+// the partition leader for the topic and return a connection to that server.
+// The original address is only used as a mechanism to discover the
+// configuration of the kafka cluster that we're connecting to.
+func (d *Dialer) DialLeader(ctx context.Context, network string, address string, topic string, partition int) (*Conn, error) {
+	b, err := d.LookupLeader(ctx, network, address, topic, partition)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := d.dialContext(ctx, network, net.JoinHostPort(b.Host, strconv.Itoa(b.Port)))
+	if err != nil {
+		return nil, err
+	}
+
+	return NewConnWith(c, ConnConfig{
+		ClientID:  d.ClientID,
+		Topic:     topic,
+		Partition: partition,
+	}), nil
+}
+
+// LookupLeader searches for the kafka broker that is the leader of the
+// partition for a given topic, returning a Broker value representing it.
+func (d *Dialer) LookupLeader(ctx context.Context, network string, address string, topic string, partition int) (Broker, error) {
+	c, err := d.DialContext(ctx, network, address)
+	if err != nil {
+		return Broker{}, err
+	}
+	defer c.Close()
+
+	brkch := make(chan Broker, 1)
+	errch := make(chan error, 1)
```

Comment:
It doesn't matter since only one value will ever be sent to either of them, and we receive at most one value as well. The code doesn't need the channels to get closed and eventually the GC collects everything that's not referenced anymore.